single-registry-architecture/README.md

```diff
@@ -0,0 +1,217 @@
+- prepare and deploy `distribution` cache instance, configure it for the new repo
+  - ensure monitoring and alerting
+- switch our CI/CD pipelines to not tag dev build artifacts
+  - use a tool (`skopeo`, `retagger`?) to replicate all (do we really need all?) the tagged versions of images we build from the old `giantswarm` repo to the new `gsoci`
+  - configure CI/CD so it uploads to both the old `giantswarm` and the new `gsoci`
+- deploy a new instance of `retagger` that will replicate all the configured public images to the new `gsoci` registry
+- switch container runtimes to use the cache as a source of images and upstream `gsoci` as a fallback
+- cleanup (after some time)
+  - delete our registries in docker hub and quay
+  - delete the old `giantswarm` registry in ACR
+- for this part, we keep China registry unchanged
+  - we reconfigure the `crsync` to replicate images from `gsoci` to `aliyun` in China
+
+#### China extension
+
+- we try to setup automated geo-replication between Europe and China
+- switch China clusters to use the new ACR-China registry
+- shut down `crsync` entirely
+- delete Aliyun registry
+
+### Helm charts
+
+The blocker for switching Helm Charts entirely to OCI is currently our artifact discovery process, which uses the
+`index.yaml` file, that is available in HTTP based Helm registries only. So, to get started, we have to switch our
+discovery process to OCI based one. Currently, we're already uploading our build artifacts to both HTTP and OCI repos,
+so nothing is blocking implementation and a switch.
+
+High level migration plan:
+
+- prepare and deploy `distribution` cache instance, configure it for the upstream ACR registry
+  - ensure monitoring and alerting
+- switch our CI/CD pipelines to not tag dev build artifacts
+  - we want to use retention policies for helm charts as well
+- implement changes in the `app-operator`
+  - use OCI discovery instead of `index.yaml` from HTTP registries
+  - (optional) solve the problem of presenting only X (by default 5) most recent versions in the ACE CR
+  - introduce a concept of primary and fall-back registry
+- stop uploading charts into HTTP registries
+- reconfigure `app-operator` to use the OCI cache as the primary registry and upstream ACR as fallback
+- cleanup HTTP registries
+- handle China as for images
+
+## Open questions
+
+Before we start on the implementation, we have to figure out answers to questions below:
+
+1. Should we use a single cache instance per installation, or rather per cloud provider + region?
+
+   The disk requirements for a cache might be significant and as a result cost of running an instance per MC can be quite
+   high in total. Additionally, we can expect many of the images to be the same for all clusters (MCs and WCs) that run
+   for the same provider. So, it seems running a single instance of the cache per cloud provider can be both much cheaper
+   and also more efficient and performant.
```

Comment:
If we only run for GS workloads I will vote for the region, it will save costs significantly.

Reply:
Where should have host the per region one? Do we loose on network / transfer costs if we do so?

Reply:
AFAIR, for example in AWS, transfer within region is free (or very cheap). Inter-region and internet is where gets more expensive.

The cache solution I've tested is a read-only cache, you can't push to it (I tried to depict it in the diagram). As for HA, it is possible, but might require multiple storage copies (for filesystem option). I guess we can avoid that by using an S3 storage, that makes the Pods almost stateless, but will be probably slower and might cost more (although it doesn't need any storage scaling, that's a pro).

Reply:
Traffic within an AWS region is 0.01 USD per GB.

Reply:
I feel my main pain point here was that this causes interconnection / dependency between MCs, making one more critical, even if the platform can fall back to upstream when cache is down.

The where do we host it issue is:
- Do we have / want to have a GS MC every region and provider our customers have?
- Would feel bad hosting in Customer A lets's say `us-west-2` and somehow expose it to Customer B `us-west-2`.

Reply:
I think we have (at least currently) 1 MC per each provider and region our customers use. We can use that to host the cache. Since the cache caches only "our" images, it should be us, who pays for that. It will still be well worth it compared to our savings on the upstream registry. As for the interconnection: I agree, that's why I want a soft-failure model: even if the cache fails, we can still use the highly performant upstream source, we just have to pay for it.

Reply:
>  who pays for that

If we running the cache in the MC account, customer will pay for it AFAIK